File: urbanairship-sdk/src/test/java/com/urbanairship/push/IncomingPushRunnableTest.java
/* Copyright 2017 Urban Airship and Contributors */

package com.urbanairship.push;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationManagerCompat;

import com.urbanairship.BaseTestCase;
import com.urbanairship.TestApplication;
import com.urbanairship.TestPushProvider;
import com.urbanairship.UAirship;
import com.urbanairship.analytics.Analytics;
import com.urbanairship.analytics.PushArrivedEvent;
import com.urbanairship.push.iam.InAppMessage;
import com.urbanairship.push.notifications.NotificationFactory;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.Shadows;
import org.robolectric.shadows.ShadowApplication;
import org.robolectric.shadows.ShadowPendingIntent;

import java.util.List;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;
import static junit.framework.Assert.assertSame;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;


public class IncomingPushRunnableTest extends BaseTestCase {

    public static final int TEST_NOTIFICATION_ID = 123;

    private Bundle pushBundle;

    private PushManager pushManager;
    private NotificationManagerCompat notificationManager;
    private Analytics analytics;

    private Notification notification;
    private NotificationFactory notificationFactory;

    private IncomingPushRunnable pushRunnable;
    private TestPushProvider testPushProvider;

    @Before
    public void setup() {
        pushBundle = new Bundle();
        pushBundle.putString(PushMessage.EXTRA_ALERT, "Test Push Alert!");
        pushBundle.putString(PushMessage.EXTRA_PUSH_ID, "testPushID");
        pushBundle.putString(PushMessage.EXTRA_SEND_ID, "testSendID");
        pushBundle.putString(PushMessage.EXTRA_NOTIFICATION_TAG, "testNotificationTag");


        pushManager = mock(PushManager.class);
        testPushProvider = new TestPushProvider();
        when(pushManager.getPushProvider()).thenReturn(testPushProvider);

        notificationManager = mock(NotificationManagerCompat.class);

        when(pushManager.isPushAvailable()).thenReturn(true);

        notification = new NotificationCompat.Builder(RuntimeEnvironment.application)
                .setContentTitle("Test NotificationBuilder Title")
                .setContentText("Test NotificationBuilder Text")
                .setAutoCancel(true)
                .build();

        notificationFactory = new NotificationFactory(TestApplication.getApplication()) {
            @Override
            public Notification createNotification(@NonNull PushMessage pushMessage, int notificationId) {
                return notification;
            }

            @Override
            public int getNextId(@NonNull PushMessage pushMessage) {
                return TEST_NOTIFICATION_ID;
            }
        };

        when(pushManager.getNotificationFactory()).thenAnswer(new Answer<Object>() {
            @Override
            public Object answer(InvocationOnMock invocation) throws Throwable {
                return notificationFactory;
            }
        });


        analytics = mock(Analytics.class);

        TestApplication.getApplication().setPushManager(pushManager);
        TestApplication.getApplication().setAnalytics(analytics);

        pushRunnable = new IncomingPushRunnable.Builder(TestApplication.getApplication())
                .setProviderClass(testPushProvider.getClass().toString())
                .setMessage(new PushMessage(pushBundle))
                .setNotificationManager(notificationManager)
                .setLongRunning(true)
                .build();
    }

    /**
     * Test displaying a notification from a push message.
     */
    @Test
    public void testDisplayNotification() {
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(true);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        pushRunnable.run();

        verify(notificationManager).notify("testNotificationTag", TEST_NOTIFICATION_ID, notification);
        verify(analytics).addEvent(any(PushArrivedEvent.class));

        ShadowPendingIntent shadowPendingIntent = Shadows.shadowOf(notification.contentIntent);
        assertTrue("The pending intent is broadcast intent.", shadowPendingIntent.isBroadcastIntent());

        Intent intent = shadowPendingIntent.getSavedIntent();
        assertEquals("The intent action should match.", intent.getAction(), PushManager.ACTION_NOTIFICATION_OPENED_PROXY);
        assertBundlesEquals("The push message bundles should match.", pushBundle, intent.getExtras().getBundle(PushManager.EXTRA_PUSH_MESSAGE_BUNDLE));
        assertEquals("One category should exist.", 1, intent.getCategories().size());
    }

    /**
     * Test receiving a push from an invalid provider.
     */
    @Test
    public void testInvalidPushProvider() {
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(true);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        pushRunnable = new IncomingPushRunnable.Builder(TestApplication.getApplication())
                .setProviderClass("wrong  class")
                .setMessage(new PushMessage(pushBundle))
                .setLongRunning(true)
                .setNotificationManager(notificationManager)
                .build();

        pushRunnable.run();

        verifyZeroInteractions(notificationManager);
    }

    /**
     * Test user notifications disabled still notifies user of a background notification.
     */
    @Test
    public void testUserNotificationsDisabled() {
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(false);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        pushRunnable.run();

        ShadowApplication shadowApplication = Shadows.shadowOf(RuntimeEnvironment.application);
        List<Intent> intents = shadowApplication.getBroadcastIntents();
        Intent i = intents.get(intents.size() - 1);
        PushMessage push = PushMessage.fromIntent(i);
        assertEquals("Intent action should be push received", i.getAction(), PushManager.ACTION_PUSH_RECEIVED);
        assertEquals("Push ID should equal pushMessage ID", "testPushID", push.getCanonicalPushId());
        assertEquals("No notification ID should be present", i.getIntExtra(PushManager.EXTRA_NOTIFICATION_ID, -1), -1);
    }

    /**
     * Test handling a background push.
     */
    @Test
    public void testBackgroundPush() {
        notification = null;
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(true);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        pushRunnable.run();

        ShadowApplication shadowApplication = Shadows.shadowOf(RuntimeEnvironment.application);
        List<Intent> intents = shadowApplication.getBroadcastIntents();
        Intent i = intents.get(intents.size() - 1);

        PushMessage push = PushMessage.fromIntent(i);
        assertEquals("Intent action should be push received", i.getAction(), PushManager.ACTION_PUSH_RECEIVED);
        assertEquals("Push ID should equal pushMessage ID", "testPushID", push.getCanonicalPushId());
        assertEquals("No notification ID should be present", -1, i.getIntExtra(PushManager.EXTRA_NOTIFICATION_ID, -1));
    }

    /**
     * Test handling an exceptions from the notification factory.
     */
    @Test
    public void testNotificationFactoryException() {
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(true);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        // Set a notification factory that throws an exception
        notificationFactory = new NotificationFactory(TestApplication.getApplication()) {
            @Override
            public Notification createNotification(@NonNull PushMessage pushMessage, int notificationId) {
                throw new RuntimeException("Unable to create and display notification.");
            }

            @Override
            public int getNextId(@NonNull PushMessage pushMessage) {
                return 0;
            }
        };

        pushRunnable.run();

        verify(notificationManager, Mockito.never()).notify(Mockito.anyInt(), any(Notification.class));
    }

    /**
     * Test notification content intent
     */
    @Test
    public void testNotificationContentIntent() {
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(true);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        PendingIntent pendingIntent = PendingIntent.getBroadcast(RuntimeEnvironment.application, 1, new Intent(), 0);
        notification = new NotificationCompat.Builder(RuntimeEnvironment.application)
                .setContentTitle("Test NotificationBuilder Title")
                .setContentText("Test NotificationBuilder Text")
                .setAutoCancel(true)
                .setContentIntent(pendingIntent)
                .build();

        pushRunnable.run();

        ShadowPendingIntent shadowPendingIntent = Shadows.shadowOf(notification.contentIntent);
        assertTrue("The pending intent is broadcast intent.", shadowPendingIntent.isBroadcastIntent());

        Intent intent = shadowPendingIntent.getSavedIntent();
        assertEquals("The intent action should match.", intent.getAction(), PushManager.ACTION_NOTIFICATION_OPENED_PROXY);
        assertEquals("One category should exist.", 1, intent.getCategories().size());
        assertNotNull("The notification content intent is not null.", pendingIntent);
        assertSame("The notification content intent matches.", pendingIntent, intent.getExtras().get(PushManager.EXTRA_NOTIFICATION_CONTENT_INTENT));
    }

    /**
     * Test notification delete intent
     */
    @Test
    public void testNotificationDeleteIntent() {
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(true);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        PendingIntent pendingIntent = PendingIntent.getBroadcast(RuntimeEnvironment.application, 1, new Intent(), 0);
        notification = new NotificationCompat.Builder(RuntimeEnvironment.application)
                .setContentTitle("Test NotificationBuilder Title")
                .setContentText("Test NotificationBuilder Text")
                .setAutoCancel(true)
                .setDeleteIntent(pendingIntent)
                .build();

        pushRunnable.run();

        ShadowPendingIntent shadowPendingIntent = Shadows.shadowOf(notification.deleteIntent);
        assertTrue("The pending intent is broadcast intent.", shadowPendingIntent.isBroadcastIntent());

        Intent intent = shadowPendingIntent.getSavedIntent();
        assertEquals("The intent action should match.", intent.getAction(), PushManager.ACTION_NOTIFICATION_DISMISSED_PROXY);
        assertEquals("One category should exist.", 1, intent.getCategories().size());
        assertNotNull("The notification delete intent is not null.", pendingIntent);
        assertSame("The notification delete intent matches.", pendingIntent, intent.getExtras().get(PushManager.EXTRA_NOTIFICATION_DELETE_INTENT));
    }

    /**
     * Test when sound is disabled the flag for DEFAULT_SOUND is removed and the notification sound
     * is set to null.
     */
    @Test
    public void testDeliverPushSoundDisabled() {
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(true);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        // Disable sound
        when(pushManager.isSoundEnabled()).thenReturn(false);

        notification.sound = Uri.parse("some://sound");
        notification.defaults = NotificationCompat.DEFAULT_ALL;

        pushRunnable.run();

        assertNull("The notification sound should be null.", notification.sound);
        assertEquals("The notification defaults should not include DEFAULT_SOUND.",
                notification.defaults & NotificationCompat.DEFAULT_SOUND, 0);
    }

    /**
     * Test when sound is disabled the flag for DEFAULT_VIBRATE is removed and the notification vibrate
     * is set to null.
     */
    @Test
    public void testDeliverPushVibrateDisabled() {
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(true);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        // Disable vibrate
        when(pushManager.isVibrateEnabled()).thenReturn(false);

        notification.defaults = NotificationCompat.DEFAULT_ALL;
        notification.vibrate = new long[] { 0L, 1L, 200L };

        pushRunnable.run();

        assertNull("The notification sound should be null.", notification.vibrate);
        assertEquals("The notification defaults should not include DEFAULT_VIBRATE.",
                notification.defaults & NotificationCompat.DEFAULT_VIBRATE, 0);
    }

    /**
     * Test delivering a push with an in-app message sets the pending notification.
     */
    @Test
    public void testDeliverPushInAppMessage() {
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(true);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        pushBundle.putString(PushMessage.EXTRA_IN_APP_MESSAGE, new InAppMessage.Builder()
                .setAlert("oh hi")
                .setExpiry(1000l)
                .create()
                .toJsonValue()
                .toString());

        pushRunnable = new IncomingPushRunnable.Builder(TestApplication.getApplication())
                .setProviderClass(testPushProvider.getClass().toString())
                .setMessage(new PushMessage(pushBundle))
                .setNotificationManager(notificationManager)
                .setLongRunning(true)
                .build();

        pushRunnable.run();

        assertEquals(new PushMessage(pushBundle).getInAppMessage(), UAirship.shared().getInAppMessageManager().getPendingMessage());
    }

    /**
     * Test the notification defaults in quiet time.
     */
    @Test
    public void testInQuietTime() {
        when(pushManager.isPushEnabled()).thenReturn(true);
        when(pushManager.isOptIn()).thenReturn(true);
        when(pushManager.isUniqueCanonicalId("testPushID")).thenReturn(true);

        when(pushManager.isVibrateEnabled()).thenReturn(true);
        when(pushManager.isSoundEnabled()).thenReturn(true);
        when(pushManager.isInQuietTime()).thenReturn(true);

        pushRunnable.run();

        assertNull("The notification sound should be null.", notification.sound);
        assertEquals("The notification defaults should not include vibrate or sound.", 0, notification.defaults);
    }



}

File: urbanairship-core/src/test/java/com/urbanairship/push/IncomingPushRunnableTest.kt
/* Copyright Airship and Contributors */
package com.urbanairship.push

import android.app.Notification
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.os.bundleOf
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.urbanairship.TestActivityMonitor
import com.urbanairship.TestApplication
import com.urbanairship.TestPushProvider
import com.urbanairship.analytics.Analytics
import com.urbanairship.job.JobDispatcher
import com.urbanairship.push.notifications.NotificationArguments
import com.urbanairship.push.notifications.NotificationArguments.Companion.newBuilder
import com.urbanairship.push.notifications.NotificationChannelCompat
import com.urbanairship.push.notifications.NotificationChannelRegistry
import com.urbanairship.push.notifications.NotificationProvider
import com.urbanairship.push.notifications.NotificationResult
import com.urbanairship.push.notifications.NotificationResult.Companion.cancel
import com.urbanairship.push.notifications.NotificationResult.Companion.notification
import com.urbanairship.push.notifications.NotificationResult.Companion.retry
import com.urbanairship.util.PendingIntentCompat
import io.mockk.Called
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RuntimeEnvironment
import org.robolectric.Shadows
import org.robolectric.annotation.Config

@RunWith(AndroidJUnit4::class)
public class IncomingPushRunnableTest {

    private var activityMonitor = TestActivityMonitor()
    private var pushBundle = bundleOf(
        PushMessage.EXTRA_ALERT to "Test Push Alert!",
        PushMessage.EXTRA_PUSH_ID to "testPushID",
        PushMessage.EXTRA_SEND_ID to "testSendID",
        PushMessage.EXTRA_NOTIFICATION_TAG to "testNotificationTag"
    )
    private var message = PushMessage(pushBundle)
    private var accengageMessage = PushMessage(
        pushBundle = bundleOf(
            "a4scontent" to "neat",
            "a4sid" to 77
        )
    )

    private var testPushProvider = TestPushProvider()
    private var mockChannelRegistry: NotificationChannelRegistry = mockk(relaxed = true)
    private var notificationProvider = TestNotificationProvider()

    private var pushManager: PushManager = mockk(relaxed = true) {
        every { pushProvider } returns testPushProvider
        every { isPushAvailable } returns true
        every { notificationChannelRegistry } returns mockChannelRegistry
        every { notificationProvider } answers { this@IncomingPushRunnableTest.notificationProvider }
    }

    private var notificationManager: NotificationManagerCompat = mockk(relaxed = true)
    private var analytics: Analytics = mockk()

    private var accengageNotificationProvider: TestNotificationProvider? = null

    private var jobDispatcher: JobDispatcher = mockk(relaxed = true)
    private lateinit var pushRunnable: IncomingPushRunnable
    private lateinit var displayRunnable: IncomingPushRunnable


    @Before
    public fun setup() {
        TestApplication.getApplication().setPushManager(pushManager)
        TestApplication.getApplication().setAnalytics(analytics)

        pushRunnable = IncomingPushRunnable.Builder(TestApplication.getApplication())
            .setProviderClass(testPushProvider.javaClass.toString())
            .setMessage(PushMessage(pushBundle))
            .setNotificationManager(notificationManager)
            .setLongRunning(true)
            .setJobDispatcher(jobDispatcher)
            .setActivityMonitor(activityMonitor)
            .build()

        displayRunnable = IncomingPushRunnable.Builder(TestApplication.getApplication())
            .setProviderClass(testPushProvider.javaClass.toString())
            .setMessage(PushMessage(pushBundle))
            .setNotificationManager(notificationManager)
            .setLongRunning(true)
            .setJobDispatcher(jobDispatcher)
            .setProcessed(true)
            .setActivityMonitor(activityMonitor)
            .build()
    }

    /**
     * Test displaying a notification from a push message.
     */
    @Test
    public fun testDisplayNotification() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        val notification = createNotification()
        notificationProvider.notification = notification
        notificationProvider.tag = "testNotificationTag"

        pushRunnable.run()

        verify { notificationManager.notify("testNotificationTag", TEST_NOTIFICATION_ID, notification) }

        val shadowPendingIntent = Shadows.shadowOf(notificationProvider.notification?.contentIntent)
        Assert.assertTrue(
            "The pending intent is an activity intent.", shadowPendingIntent.isActivity
        )

        val intent = shadowPendingIntent.savedIntent
        Assert.assertEquals(
            "The intent action should match.",
            intent.action,
            PushManager.ACTION_NOTIFICATION_RESPONSE
        )
        Assert.assertEquals(
            "The push message bundles should match.",
            pushBundle,
            intent.extras?.getBundle(PushManager.EXTRA_PUSH_MESSAGE_BUNDLE)
        )
        Assert.assertEquals("One category should exist.", 1, intent.categories.size)

        verify { pushManager.onPushReceived(message, true) }
        verify { pushManager.onNotificationPosted(message, TEST_NOTIFICATION_ID, "testNotificationTag") }
    }

    /**
     * Test ignoring push from other vendors.
     */
    @Test
    public fun test() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        notificationProvider.notification = createNotification()
        notificationProvider.tag = "testNotificationTag"

        pushRunnable = IncomingPushRunnable.Builder(TestApplication.getApplication())
            .setProviderClass("wrong  class")
            .setMessage(PushMessage(pushBundle))
            .setLongRunning(true)
            .setNotificationManager(notificationManager)
            .build()

        pushRunnable.run()

        verify { notificationManager wasNot Called }
    }

    /**
     * Test receiving a push from an invalid provider.
     */
    @Test
    public fun testInvalidPushProvider() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        pushRunnable = IncomingPushRunnable.Builder(TestApplication.getApplication())
            .setProviderClass("wrong  class")
            .setMessage(PushMessage(pushBundle))
            .setLongRunning(true)
            .setNotificationManager(notificationManager)
            .build()

        pushRunnable.run()

        verify { notificationManager wasNot Called }
    }

    /**
     * Test user notifications disabled still notifies user of a background notification.
     */
    @Test
    public fun testUserNotificationsDisabled() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns false
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        pushRunnable.run()

        verify { pushManager.onPushReceived(message, false) }
    }

    /**
     * Test suppress message in foreground when isForegroundDisplayable is false.
     */
    @Test
    public fun testNotForegroundDisplayable() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        activityMonitor.foreground()

        notificationProvider.notification = createNotification()
        notificationProvider.tag = "testNotificationTag"

        pushBundle.putString("com.urbanairship.foreground_display", "false")
        message = PushMessage(pushBundle)

        pushRunnable = IncomingPushRunnable.Builder(TestApplication.getApplication())
            .setProviderClass(testPushProvider.javaClass.toString())
            .setMessage(message)
            .setNotificationManager(notificationManager)
            .setLongRunning(true)
            .setJobDispatcher(jobDispatcher)
            .setActivityMonitor(activityMonitor)
            .build()

        pushRunnable.run()

        verify { pushManager.onPushReceived(message, false) }
    }

    /**
     * Test message in foreground when isForegroundDisplayable is true.
     */
    @Test
    public fun testForegroundDisplayable() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true
        every { pushManager.foregroundNotificationDisplayPredicate } returns null

        activityMonitor.foreground()

        notificationProvider.notification = createNotification()
        notificationProvider.tag = "testNotificationTag"

        pushBundle.putString("com.urbanairship.foreground_display", "true")
        message = PushMessage(pushBundle)

        pushRunnable = IncomingPushRunnable.Builder(TestApplication.getApplication())
            .setProviderClass(testPushProvider.javaClass.toString())
            .setMessage(message)
            .setNotificationManager(notificationManager)
            .setLongRunning(true)
            .setJobDispatcher(jobDispatcher)
            .setActivityMonitor(activityMonitor)
            .build()

        pushRunnable.run()

        verify { pushManager.onPushReceived(message, true) }
    }

    @Test
    public fun testForegroundDisplayPredicate() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true
        every { pushManager.foregroundNotificationDisplayPredicate } returns { false }

        activityMonitor.foreground()

        notificationProvider.notification = createNotification()
        notificationProvider.tag = "testNotificationTag"
        message = PushMessage(pushBundle)

        pushRunnable = IncomingPushRunnable.Builder(TestApplication.getApplication())
            .setProviderClass(testPushProvider.javaClass.toString())
            .setMessage(message)
            .setNotificationManager(notificationManager)
            .setLongRunning(true)
            .setJobDispatcher(jobDispatcher)
            .setActivityMonitor(activityMonitor)
            .build()

        pushRunnable.run()

        verify { pushManager.onPushReceived(message, false) }
    }

    /**
     * Test handling a background push.
     */
    @Test
    public fun testBackgroundPush() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        pushRunnable.run()

        verify { pushManager.onPushReceived(message, false) }
    }

    /**
     * Test handling an exceptions from the notification provider.
     */
    @Test
    public fun testNotificationProviderException() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        // Set a notification factory that throws an exception
        notificationProvider = object : TestNotificationProvider() {
            override fun onCreateNotification(
                context: Context, arguments: NotificationArguments
            ): NotificationResult {
                throw RuntimeException("Unable to create and display notification.")
            }
        }

        pushRunnable.run()

        verify(exactly = 0) { notificationManager.notify(any(), any(), any()) }
        verify(exactly = 0) { jobDispatcher.dispatch(any()) }
    }

    @Test
    public fun testNotificationProviderSuccess() {
        val notification = createNotification()
        notificationProvider.notification = notification

        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        pushRunnable.run()

        verify { notificationManager.notify(null, TEST_NOTIFICATION_ID, notification)}
        verify(exactly = 0) { jobDispatcher.dispatch(any()) }
    }

    /**
     * Test that when the provider returns a cancel status, no notification is posted and no jobs are scheduled
     */
    @Test
    public fun testNotificationProviderResultCancel() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        notificationProvider = object : TestNotificationProvider() {
            override fun onCreateNotification(
                context: Context, arguments: NotificationArguments
            ): NotificationResult {
                return cancel()
            }
        }

        displayRunnable.run()

        verify(exactly = 0) { notificationManager.notify(any(), any(), any()) }
        verify(exactly = 0) { jobDispatcher.dispatch(any()) }
    }

    /**
     * Test that when the factory returns a retry status, no notification is posted and a retry job is scheduled
     */
    @Test
    public fun testNotificationFactoryResultRetry() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        notificationProvider = object : TestNotificationProvider() {
            override fun onCreateNotification(
                context: Context, arguments: NotificationArguments
            ): NotificationResult {
                return retry()
            }
        }

        displayRunnable.run()

        verify(exactly = 0) { notificationManager.notify(any(), any(), any()) }
        verify { jobDispatcher.dispatch(any()) }
    }

    /**
     * Test that when the factory returns a successful result, a notification is posted and no jobs are scheduled.
     */
    @Test
    public fun testNotificationFactoryResultOK() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        val notification = createNotification()
        notificationProvider.notification = notification

        displayRunnable.run()

        verify { notificationManager.notify(null, TEST_NOTIFICATION_ID, notification) }
        verify(exactly = 0) { jobDispatcher.dispatch(any()) }
    }

    /**
     * Test notification content intent
     */
    @Test
    public fun testNotificationContentIntent() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        val pendingIntent = PendingIntentCompat
            .getBroadcast(ApplicationProvider.getApplicationContext(), 1, Intent(), 0)
        notificationProvider.notification = createNotification()
        notificationProvider.tag = "cool-tag"
        notificationProvider.notification!!.contentIntent = pendingIntent

        pushRunnable.run()

        val shadowPendingIntent =
            Shadows.shadowOf(notificationProvider.notification?.contentIntent)
        Assert.assertTrue(
            "The pending intent is an activity intent.", shadowPendingIntent.isActivityIntent
        )

        val intent = shadowPendingIntent.savedIntent
        Assert.assertEquals(
            "The intent action should match.",
            intent.action,
            PushManager.ACTION_NOTIFICATION_RESPONSE
        )
        Assert.assertEquals("One category should exist.", 1, intent.categories.size)
        Assert.assertNotNull("The notification content intent is not null.", pendingIntent)
        Assert.assertSame(
            "The notification content intent matches.",
            pendingIntent,
            intent.extras?.get(PushManager.EXTRA_NOTIFICATION_CONTENT_INTENT)
        )
        Assert.assertSame(
            "cool-tag", intent.extras?.getString(PushManager.EXTRA_NOTIFICATION_TAG)
        )
        Assert.assertSame(
            TEST_NOTIFICATION_ID, intent.extras?.getInt(PushManager.EXTRA_NOTIFICATION_ID)
        )
    }

    /**
     * Test notification delete intent
     */
    @Test
    public fun testNotificationDeleteIntent() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        val pendingIntent = PendingIntentCompat
            .getBroadcast(RuntimeEnvironment.application, 1, Intent(), 0)
        notificationProvider.notification = createNotification()
        notificationProvider.notification?.deleteIntent = pendingIntent

        pushRunnable.run()

        val shadowPendingIntent =
            Shadows.shadowOf(notificationProvider.notification?.deleteIntent)
        Assert.assertTrue(
            "The pending intent is broadcast intent.", shadowPendingIntent.isBroadcast
        )

        val intent = shadowPendingIntent.savedIntent
        Assert.assertEquals(
            "The intent action should match.",
            intent.action,
            PushManager.ACTION_NOTIFICATION_DISMISSED
        )
        Assert.assertEquals("One category should exist.", 1, intent.categories.size)
        Assert.assertNotNull("The notification delete intent is not null.", pendingIntent)
        Assert.assertSame(
            "The notification delete intent matches.",
            pendingIntent,
            intent.extras?.get(PushManager.EXTRA_NOTIFICATION_DELETE_INTENT)
        )
    }

    /**
     * Test that when a push is delivered pre-Oreo the notification settings are drawn
     * from our notification channel compat layer.
     */
    @Test
    @Config(sdk = [25])
    public fun testDeliverPushPreOreo() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        // Create a channel and set some non-default values
        val channelCompat = NotificationChannelCompat(
            TEST_NOTIFICATION_CHANNEL_ID,
            "Test Notification Channel",
            NotificationManager.IMPORTANCE_HIGH
        )
        channelCompat.sound = Uri.parse("cool://sound")
        channelCompat.enableVibration(true)
        channelCompat.enableLights(true)
        channelCompat.lightColor = 123

        every { mockChannelRegistry.getNotificationChannelSync(TEST_NOTIFICATION_CHANNEL_ID) } returns channelCompat
        notificationProvider.notification = createNotification()

        pushRunnable.run()

        val notification = notificationProvider.notification

        Assert.assertEquals(notification!!.sound, channelCompat.sound)
        Assert.assertEquals(NotificationManager.IMPORTANCE_HIGH, channelCompat.importance)
        Assert.assertEquals(
            notification.defaults and Notification.DEFAULT_VIBRATE, Notification.DEFAULT_VIBRATE
        )
        Assert.assertEquals(notification.ledARGB, channelCompat.lightColor)
    }

    /**
     * Test remote data notifications
     */
    @Test
    public fun testRemoteDataMessage() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        // Set a notification factory that throws an exception
        notificationProvider = mockk()

        val pushBundle = bundleOf(
            PushMessage.EXTRA_PUSH_ID to "testPushID",
            PushMessage.EXTRA_SEND_ID to "testSendID",
            PushMessage.REMOTE_DATA_UPDATE_KEY to "true"
        )

        val message = PushMessage(pushBundle)

        IncomingPushRunnable.Builder(TestApplication.getApplication())
            .setProviderClass(testPushProvider.javaClass.toString())
            .setMessage(message)
            .setNotificationManager(notificationManager)
            .setLongRunning(true)
            .setJobDispatcher(jobDispatcher)
            .build()
            .run()

        verify(exactly = 0) { notificationManager.notify(any(), any(), any()) }
        verify(exactly = 0) { jobDispatcher.dispatch(any()) }
        verify { notificationProvider wasNot Called }
        verify { pushManager.onPushReceived(message, false) }
    }

    @Test
    public fun testNullNotificationChannel() {
        every { pushManager.isPushEnabled } returns true
        every { pushManager.isOptIn } returns true
        every { pushManager.isUniqueCanonicalId("testPushID") } returns true

        notificationProvider.notification = NotificationCompat.Builder(TestApplication.getApplication())
                .setContentTitle("Test NotificationBuilder Title")
                .setContentText("Test NotificationBuilder Text")
                .setAutoCancel(true)
                .build()

        notificationProvider.tag = "testNotificationTag"

        pushRunnable.run()

        verify { mockChannelRegistry wasNot Called }
        verify { notificationManager.notify("testNotificationTag", TEST_NOTIFICATION_ID, notificationProvider.notification!!) }
        verify { pushManager.onPushReceived(message, true) }
        verify { pushManager.onNotificationPosted(message, TEST_NOTIFICATION_ID, "testNotificationTag") }
    }

    private fun createNotification(): Notification = NotificationCompat
        .Builder(TestApplication.getApplication(), "some-channel")
        .setContentTitle("Test NotificationBuilder Title")
        .setContentText("Test NotificationBuilder Text")
        .setAutoCancel(true)
        .build()

    public open class TestNotificationProvider : NotificationProvider {

        public var notification: Notification? = null
        public var tag: String? = null

        override fun onCreateNotificationArguments(
            context: Context, message: PushMessage
        ): NotificationArguments {
            return newBuilder(message).setNotificationChannelId(TEST_NOTIFICATION_CHANNEL_ID)
                .setNotificationId(tag, TEST_NOTIFICATION_ID).build()
        }

        override fun onCreateNotification(
            context: Context, arguments: NotificationArguments
        ): NotificationResult {
            return if (notification != null) {
                notification(
                    notification!!
                )
            } else {
                cancel()
            }
        }

        override fun onNotificationCreated(
            context: Context, notification: Notification, arguments: NotificationArguments
        ) {
        }
    }

    public companion object {

        private const val TEST_NOTIFICATION_ID = 123
        private const val TEST_NOTIFICATION_CHANNEL_ID = "Test notification channel"
    }
}
